Python 3 error: step2_exchange confusing unicode/bytes

happyspace opened the pull request:
In python 3, the code received from the browser will be binary instead of a string. Expand the API for oauth_flow.step2_exchange(code) to allow this value to be passed in as is rather than decoding.

for example:
credentials = self.flow.step2_exchange(b'some random code')

test:
tox -e py34  -- tests.test_client:OAuth2WebServerFlowTest.test_exchange_success_binary_code

resolves:  #443

nathanielmanistaatgoogle commented on tests/test_client.py:
```diff
@@ -1201,6 +1201,21 @@ def test_exchange_success(self):
+        http = HttpMockSequence([({'status': '200'}, payload)])
+        credentials = self.flow.step2_exchange(b'some random code', http=http)
+        self.assertEqual('SlAV32hkKG', credentials.access_token)
+        self.assertNotEqual(None, credentials.token_expiry)
```
`self.assertIsNotNone` rather than comparing with None for equality, right?

dhermes commented:
LGTM pending @nathanielmanistaatgoogle comment

nathanielmanistaatgoogle commented on tests/test_client.py:
```diff
@@ -1201,6 +1201,21 @@ def test_exchange_success(self):
 
+    def test_exchange_success_binary_code(self):
+        payload = (b'{'
+                   b'  "access_token":"SlAV32hkKG",'
```
Please don't repeat the literal values of your test data. Please instead create local constants for your test data values and then use those constants both where you supply the test data to the system under test and where you confirm the correctness of the system under test.

I recognize that this is asking you not to follow the pattern of the other test methods nearby.

happyspace replied:
Wonderful thanks! Yes guilty as charged, I used the test_exchange_test as the basis for this test. Could you point me to a good example test?

nathanielmanistaatgoogle replied:
Uhh...

... after some searching: [`ServiceAccountCredentialsTest.test_from_json_keyfile_name_factory`](https://github.com/google/oauth2client/blob/b605997beaa6361ab0811ffe62dfaf8de0eed03e/tests/test_service_account.py#L110). Although even there I'd object to the literal strings "foo" and "bar" each appearing more than once in the test data set.

But do you see what I mean about method-local constants, so that the literal values used as the test inputs are only written once?

happyspace replied:
Wonderful yes!

nathanielmanistaatgoogle commented:
Please amend your commit message to conform to [these guidelines](http://chris.beams.io/posts/git-commit/#seven-rules).

nathanielmanistaatgoogle commented:
Closed?

happyspace committed:
Expand API for step2_exchange

In python 3, the code received from the browser will be binary instead of a string. Expand the API for oauth_flow.step2_exchange(code) to allow this value to be passed as is rather than decoding.

for example:
credentials = self.flow.step2_exchange(b'some random code')

test:
tox -e py34  -- tests.test_client:OAuth2WebServerFlowTest.test_exchange_success_binary_code

resolves:  #443, #446

happyspace commented:
Sorry I was updating the comment and wiped everything out with a force push. grrr...
May need to redo the pull request?

nathanielmanistaatgoogle commented on tests/test_client.py:
```diff
@@ -1201,6 +1202,26 @@ def test_exchange_success(self):
+        http = HttpMockSequence([({'status': '200'}, _to_bytes(payload))])
+        credentials = self.flow.step2_exchange(binary_code, http=http)
+        self.assertEqual(access_token, credentials.access_token)
+        self.assertNotEqual(None, credentials.token_expiry)
```
`self.assertNotNone`, please. Right?

nathanielmanistaatgoogle commented:
Please squash commits with each round of code review back-and-forth. Are you sure that your commit message summary line is [fifty characters or fewer](http://chris.beams.io/posts/git-commit/#seven-rules)?
